plugin/pkg/scheduler/core/generic_scheduler.go

```diff
@@ -424,6 +445,115 @@ func EqualPriorityMap(_ *v1.Pod, _ interface{}, nodeInfo *schedulercache.NodeInf
+// NOTE: This function assumes that it is never called if "pod" cannot be scheduled
+// due to pod affinity, node affinity, or node anti-affinity reasons. None of
+// these predicates can be satisfied by removing more pods from the node.
+func selectVictimsOnNode(pod *v1.Pod, meta interface{}, nodeInfo *schedulercache.NodeInfo, fitPredicates map[string]algorithm.FitPredicate) ([]*v1.Pod, bool) {
+	higherPriority := func(pod1, pod2 interface{}) bool {
+		return util.GetPodPriority(pod1.(*v1.Pod)) > util.GetPodPriority(pod2.(*v1.Pod))
+	}
+	potentialVictims := util.SortableList{CompFunc: higherPriority}
+	nodeInfoCopy := nodeInfo.Clone()
+	// As the first step, remove all the lower priority pods from the node and
+	// check if the given pod can be scheduled.
+	podPriority := util.GetPodPriority(pod)
+	for _, p := range nodeInfoCopy.Pods() {
+		if util.GetPodPriority(p) < podPriority {
+			potentialVictims.Items = append(potentialVictims.Items, p)
+			nodeInfoCopy.RemovePod(p)
+		}
+	}
+	potentialVictims.Sort()
+	// If the new pod does not fit after removing all the lower priority pods,
+	// we are almost done and this node is not suitable for preemption. The only condition
+	// that we should check is if the "pod" is failing to schedule due to pod affinity
+	// failure.
+	if fits, failedPredicates, err := podFitsOnNode(pod, meta, nodeInfoCopy, fitPredicates, nil); !fits {
+		if err != nil {
+			glog.Warningf("Encountered error while selecting victims on node %v: %v", nodeInfo.Node().Name, err)
+			return nil, false
+		}
+		// If the new pod still cannot be scheduled for any reason other than pod
+		// affinity, the new pod will not fit on this node and we are done here.
+		affinity := pod.Spec.Affinity
+		if affinity == nil || affinity.PodAffinity == nil {
+			return nil, false
+		}
+		for _, failedPred := range failedPredicates {
+			if failedPred != predicates.ErrPodAffinityNotMatch {
+				return nil, false
+			}
+		}
+		// If we reach here, it means that the pod cannot be scheduled only due to
+		// pod affinity. Let's try adding pods one at a time and see if any of them
+		// satisfies the pod affinity rules.
+		for i, p := range potentialVictims.Items {
+			existingPod := p.(*v1.Pod)
+			nodeInfoCopy.AddPod(existingPod)
+			if fits, _, _ = podFitsOnNode(pod, meta, nodeInfoCopy, fitPredicates, nil); !fits {
+				nodeInfoCopy.RemovePod(existingPod)
+			} else {
+				// We found the pod needed to satisfy pod affinity. Let's remove it from
+				// potential victims list.
+				// NOTE: We assume that pod affinity can be satisfied by only one pod,
+				// not multiple pods. This is how scheduler works today.
+				potentialVictims.Items = append(potentialVictims.Items[:i], potentialVictims.Items[i+1:]...)
+				break
+			}
+		}
+		if !fits {
+			return nil, false
+		}
+	}
+	victims := []*v1.Pod{}
+	// Try to reprieve as may pods as possible starting from the highest priority one.
```

Comment:
This problem exists, but it exists even today as Kubelet may evict a pod due to resource pressure and the eviction could violate affinity rules.
For preemption, it will be computationally very expensive to take care of affinity as topology key may be much broad than a single node, for example, it could be "region". In that case many pods must be checked.
A user can prevent this scenario by having the same priority for pods that have affinity on one another.